### PR TITLE
Initial Classes

### PR DESCRIPTION
**Pull request by mecha**
The initial classes. Or more specifically, traits.

This PR introduces **four** main traits for PDO based resource models, as well as various other traits that each provide a fragment of the implementation of the query building process.

1. [`PdoSelectCapableTrait`]
2. [`PdoInsertCapableTrait`]
3. [`PdoUpdateCapableTrait`]
4. [`PdoDeleteCapableTrait`]

[`AbstractPdoResourceModel`] is also provided for basic PDO awareness and query execution functionality.

[`PdoSelectCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoSelectCapableTrait.php
[`PdoInsertCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoInsertCapableTrait.php
[`PdoUpdateCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoUpdateCapableTrait.php
[`PdoDeleteCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoDeleteCapableTrait.php

[`AbstractPdoResourceModel`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/AbstractPdoResourceModel.php

### CODE REVIEW

File: src/Pdo/PdoAwareTrait.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                null,
+                $pdo
+            );
+        }
+
+        $this->pdo = $pdo;
+    }
+
+    /**
+     * Creates a new Dhii invalid argument exception.
```
Comment (XedinUnknown) [inline src/Pdo/PdoAwareTrait.php]: Remove the "Dhii".

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * Common functionality for objects that can built INSERT SQL queries.
+ *
+ * @since [*next-version*]
+ */
+trait BuildInsertSqlCapableTrait
+{
+    /**
+     * Builds a INSERT SQL query.
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: "a" -> "an".

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 3]
```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace RebelCode\Storage\Resource\Pdo\Query;
+
+use Dhii\Util\String\StringableInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+
+/**
+ * Common functionality for objects that can built INSERT SQL queries.
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: "built" -> "build"

File: src/Pdo/Query/BuildDeleteSqlCapableTrait.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+     * @param LogicalExpressionInterface|null $condition    Optional condition that records must satisfy to be deleted.
+     * @param string[]|Stringable[]           $columnMap    Optional mapping of field names to column names.
+     * @param string[]|Stringable[]           $valueHashMap Optional mapping of term names to their hashes
+     *
+     * @return string The built DELETE query.
+     */
+    protected function _buildDeleteSql(
+        $table,
+        LogicalExpressionInterface $condition = null,
+        array $columnMap = [],
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildDeleteSqlCapableTrait.php]: Why is it the concern of these methods to deal with field-to-column mapping?

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+     * @param array                 $columnMap    Optional mapping of field names to column names.
+     * @param array                 $valueHashMap Optional map of value names and their hashes.
+     *
+     * @return string The built INSERT query.
+     */
+    protected function _buildInsertSql(
+        $table,
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: Why is it the concern of these methods to deal with field-to-column mapping?

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+     *                                            values.
+     * @param array                 $columnMap    Optional mapping of field names to column names.
+     * @param array                 $valueHashMap Optional map of value names and their hashes.
+     *
+     * @return string The built VALUES list or an empty string if the row set has no entries.
+     */
+    protected function _buildSqlValuesList(
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: Why is it the concern of these methods to deal with field-to-column mapping?

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            return '';
+        }
+
+        $values = [];
+
+        foreach ($rowSet as $_rowData) {
+            $values[] = $this->_buildSqlRowValues($columns, $_rowData, $columnMap, $valueHashMap);
+        }
+
+        return sprintf('VALUES %s', implode(', ', $values));
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: What do you think about leaving the `VALUES` word itself up to the consumer of this method?

File: src/Pdo/Query/BuildInsertSqlCapableTrait.php [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+     * @param array $rowData      The row data, as a map of column names to row values.
+     * @param array $columnMap    Optional mapping of field names to column names.
+     * @param array $valueHashMap Optional map of value names and their hashes.
+     *
+     * @return string The build row values as a comma separated list in parenthesis.
+     */
+    protected function _buildSqlRowValues(
+        array $columns,
+        array $rowData,
+        array $columnMap = [],
```
Comment (XedinUnknown) [inline src/Pdo/Query/BuildInsertSqlCapableTrait.php]: Why is it the concern of these methods to deal with field-to-column mapping?

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

**Comment by XedinUnknown** [comment]
1. Column map is an implementation detail. Furthermore, it's a configuration-time detail, as opposed to call-time. This is because the CQRS model knows the field-to-table mapping.
2. The expression field-to-field mapping is a different story. This can be done when building the expression, or when un-serializing a serialized expression. Eventual expressions must have the fields be final. They can later be translated by the CQRS model into a column. Therefore, the column map should no longer be accepted as method arg, but instead the method can retrieve it via a different abstract method.
3. To facilitate this translation, the column map needs to be augmented to be able to use an object as value. This object provides the column name and table name both. For example, `$map = ['age' => new EntityField('user', 'age')]`.

Eventually, the hash mapping remains abstract but becomes more concrete specifically for the PDO trait model. At the same time, the field-to-column mapping remains abstract for *any* SQL trait model, because they may chose to map or not map.

**Comment by mecha** [comment]
@XedinUnknown Take a look now. I've removed all uses of `$columnMap` parameters. The `RenderSqlConditionCapableTrait` now depends on an abstract `_getFieldColumnMap()` method. Furthermore, it expects it to return a map of [`EntityFieldInterface`](https://github.com/Dhii/sql-interface/blob/task/initial-interfaces/src/EntityFieldInterface.php) instances.

File: src/Pdo/PdoDeleteCapableTrait.php [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    abstract protected function _getSqlDeleteTable();
+
+    /**
+     * Retrieves the SQL DELETE query column "field" names.
+     *
+     * @since [*next-version*]
+     *
+     * @return string[]|Stringable[] A list of field names.
+     */
+    abstract protected function _getSqlDeleteFieldNames();
```
Comment (XedinUnknown) [inline src/Pdo/PdoDeleteCapableTrait.php:77]: This method returns a map, but the name doesn't hint that.

File: src/Pdo/PdoDeleteCapableTrait.php [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * @since [*next-version*]
+     *
+     * @param LogicalExpressionInterface|null $condition Optional condition that records must satisfy to be deleted.
+     *
+     * @return PDOStatement The executed PDO statement.
+     */
+    protected function _delete(LogicalExpressionInterface $condition = null)
+    {
+        $fieldNames = $this->_getSqlDeleteFieldNames();
```
Comment (XedinUnknown) [inline src/Pdo/PdoDeleteCapableTrait.php:28]: This variable holds a map, but the name doesn't hint that.

File: src/Pdo/PdoInsertCapableTrait.php [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    /**
+     * Extracts and retrieves the data hash map from record data.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $record An associative array mapping field names to their values for a single record.
+     *
+     * @return array The generated value hash map, mapping values to their respective hashes.
```
Comment (XedinUnknown) [inline src/Pdo/PdoInsertCapableTrait.php]: I think the return type could be more narrow and descriptive.

File: src/Pdo/PdoInsertCapableTrait.php [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+ * @since [*next-version*]
+ */
+trait PdoInsertCapableTrait
+{
+    /**
+     * Executes an INSERT SQL query, inserting records into the database.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $records A list of containers, each containing the record data to insert.
```
Comment (XedinUnknown) [inline src/Pdo/PdoInsertCapableTrait.php]: Why not make this a `Traversable`?

File: src/Pdo/PdoInsertCapableTrait.php [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+        return $statement;
+    }
+
+    /**
+     * Extracts and retrieves the data hash map from record data.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $record An associative array mapping field names to their values for a single record.
```
Comment (XedinUnknown) [inline src/Pdo/PdoInsertCapableTrait.php]: How about a `Traversable`?

File: src/Pdo/PdoInsertCapableTrait.php [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    abstract protected function _getPdoValueHashString($value);
+
+    /**
+     * Builds an INSERT SQL query.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable     $table        The name of the table to insert into.
+     * @param string[]|Stringable[] $columns      A list of columns names. The order is preserved in the built query.
+     * @param array                 $rowSet       The record data as a map of column names to values.
```
Comment (XedinUnknown) [inline src/Pdo/PdoInsertCapableTrait.php:163]: `Traversable` of containers?

File: src/Pdo/PdoInsertCapableTrait.php (hunk 11 above)
Comment (XedinUnknown) [reply src/Pdo/PdoInsertCapableTrait.php]: How about I create a new `MapInterface extends ContainerInterface, Traversable`, and you can return that?

File: src/Pdo/PdoSelectCapableTrait.php [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+        return $this->_executePdoQuery($query, $valueHashMap)->fetchAll(PDO::FETCH_ASSOC);
+    }
+
+    /**
+     * Builds a SELECT SQL query.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[]           $columns        A list of names of columns to select.
```
Comment (XedinUnknown) [inline src/Pdo/PdoSelectCapableTrait.php:57]: In such cases, should we perhaps also allow `Traversable`?

File: src/Pdo/PdoUpdateCapableTrait.php [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+ * @since [*next-version*]
+ */
+trait PdoUpdateCapableTrait
+{
+    /**
+     * Executes an UPDATE SQL query, updating records in the database that satisfy the given condition.
+     *
+     * @since [*next-version*]
+     *
+     * @param array                           $changeSet A map of field names mapping to the values to change.
```
Comment (XedinUnknown) [inline src/Pdo/PdoUpdateCapableTrait.php:24]: Can these be SQL expressions? What happens if they are? What is this map expected to contain?

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
I totally love this implementation. It looks like a lot of work, thought, and normalization went in. Great job!

In terms of improvement, I think it could be a good idea to at least optionally allow containers or iterables - depending on how a method uses the parameter. I will make a new interface for maps that is an iterable container, and methods can create those.

A few minor naming changes could also be a good idea. Maybe even inline comments: I found it hard to follow some of the primary methods for CRUD.

In terms of documentation, I would like to see list/map parameter docs explain what the list/map is supposed to contain.

**Comment by XedinUnknown** [comment]
Please add branch alias for `develop`?

File: src/Pdo/PdoDeleteCapableTrait.php (hunk 10 above)
Comment (mecha) [reply src/Pdo/PdoDeleteCapableTrait.php:28]: It doesn't hold a map. It holds a list of field names.

File: src/Pdo/PdoDeleteCapableTrait.php (hunk 9 above)
Comment (mecha) [reply src/Pdo/PdoDeleteCapableTrait.php:77]: It doesn't return a map. It returns a list of field names.

File: src/Pdo/PdoInsertCapableTrait.php (hunk 12 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: Don't know why not, it's a good idea. I'll make the changes.

File: src/Pdo/PdoInsertCapableTrait.php (hunk 13 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: Don't know why not, it's a good idea. I'll make the changes.

File: src/Pdo/PdoInsertCapableTrait.php (hunk 14 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php:163]: Don't think it's possible. The input set of record data is a mapping of **field names** to values, but that needs to be transformed into a map of **column names** names to values. Unless of course you mean the input can be a `Traversable` of containers but the output remains an array.

Of course, there's also the argument of "adapting from the outside vs adapting internally".

File: src/Pdo/PdoInsertCapableTrait.php (hunk 11 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: I don't think this is an adequate place to be dealing with high-level data structures. This is not API-level, but implementation logic code, that will feed its result into another non-API method.

Thoughts?

FYI this method still needs to return an array as it's required to be used in PHP array functions later.

File: src/Pdo/PdoSelectCapableTrait.php (hunk 15 above)
Comment (mecha) [reply src/Pdo/PdoSelectCapableTrait.php:57]: Yes, good idea. `Traversable` fits with the way this parameter is being used.

File: src/Pdo/PdoUpdateCapableTrait.php (hunk 16 above)
Comment (mecha) [reply src/Pdo/PdoUpdateCapableTrait.php:24]: > Can these be SQL expressions?

I guess they can be. I considered this for a while, but I hadn't found whether expressions are supported in UPDATE queries or whether it was right to accept them given that only a few SQL databases/dialects might support them.

> What happens if they are?

This file would have more characters.

> What is this map expected to contain?

Field names mapping to new values. These are used to build the `SET` portion of the query.

```
UPDATE my_table SET fieldName = "value", otherField = "other value";
```

File: src/Pdo/PdoInsertCapableTrait.php (hunk 14 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php:163]: As I write this I realize that no field-to-column transformation is being done for the `$records`, but there should be.

File: src/Pdo/PdoInsertCapableTrait.php (hunk 12 above)
Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: After giving it some more thought, it makes sense to be able to accept a `Traversable` of `ContainerInterface`s here, as you suggested [here](https://github.com/RebelCode/cqrs-resource-model-abstract/pull/1#discussion_r158841594). Not only does this make sense given how we view container, but it would also allow us to pass data models directly without having to transform them into arrays first.

Alternatively, we can also accept a `Traversable` of `Traversable`s.

Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: I would say we pick something and stick to it, at least for each record, i.e. an array of traversable of _either_ sub-arrays or containers.

At this point, the only thing that's making me hesitant from using containers is the fact that consumers are then bound to using them which might mean they would have to construct them. Is this a valid concern?

Comment (mecha) [reply src/Pdo/PdoInsertCapableTrait.php]: @XedinUnknown seems to have identified the same problem while working on something else, and has come up with a solution which I quite like. It involves depending on abstract methods which functional similar to a container's `get()` and `has()` methods, except that they have a container/array as parameter.

This means that our classes can use these methods to retrieve something by key while not knowing what they are retrieving from, be it a container, array or something else. But most importantly, it allows us to throw container-related exceptions for non-container collections, such as arrays. With traits that implement this functionality, this seems to be the best approach.

So I will make the resource model abstracts accept both arrays and container, and depend on these abstract methods for internal reading.

File: src/Pdo/PdoDeleteCapableTrait.php (hunk 10 above)
Comment (XedinUnknown) [reply src/Pdo/PdoDeleteCapableTrait.php:28]: In this case, we specifically need a list of *unique* field names. Although a map probably exists somewhere else, it is not usable here.

**Review by XedinUnknown — APPROVED** [review]
